presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFiltersCollector.java

```diff
@@ -13,28 +13,52 @@
  */
 package io.prestosql.sql.planner;
 
+import com.google.common.util.concurrent.SettableFuture;
+import io.airlift.log.Logger;
 import io.prestosql.spi.predicate.Domain;
 import io.prestosql.spi.predicate.TupleDomain;
+import io.prestosql.sql.planner.plan.DynamicFilterId;
 
 import javax.annotation.concurrent.GuardedBy;
 import javax.annotation.concurrent.ThreadSafe;
 
+import java.util.Collection;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
+import java.util.Objects;
 import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
+import static com.google.common.base.Verify.verify;
 import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static io.airlift.concurrent.MoreFutures.addSuccessCallback;
+import static io.airlift.concurrent.MoreFutures.toCompletableFuture;
+import static io.airlift.concurrent.MoreFutures.whenAnyComplete;
+import static io.prestosql.spi.connector.DynamicFilter.NOT_BLOCKED;
 
 @ThreadSafe
 class LocalDynamicFiltersCollector
 {
+    private static final Logger log = Logger.get(LocalDynamicFiltersCollector.class);
+
     /**
      * May contains domains for dynamic filters for different table scans
      * (e.g. in case of co-located joins).
      */
     @GuardedBy("this")
     private Map<Symbol, Domain> dynamicFilterDomainsResult = new HashMap<>();
 
+    // Each future blocks until its dynamic filter is collected.
+    @GuardedBy("this")
+    private Map<DynamicFilterId, SettableFuture<Void>> futures = new HashMap<>();
```

Comment:
This becomes very similar to `DynamicFilterService` from https://github.com/prestosql/presto/pull/4224.
Can we untangle these classes a bit?

* `LocalDynamicFiltersCollector` should hold `Map<DynamicFilterId, SettableFuture<Domain>> dynamicFilters`

* There should be method `DynamicFilter LocalDynamicFiltersCollector#getDynamicFilter(List<DynamicFilters.Descriptor> dynamicFilters, Map<Symbol, ColumnHandle> columnHandles)`, similar as in `DynamicFilterService`

* `LocalDynamicFiltersCollector#addDynamicFilter(Map<Symbol, Domain> dynamicFilterDomains)` would become `addDynamicFilter(Map<DynamicFilterId, Domain> dynamicFilterDomains)`

This would simplify `LocalDynamicFilterConsumer` as it wouldn't need to perform `dynamic filter id -> symbol` translation in `io.prestosql.sql.planner.LocalDynamicFilterConsumer#convertTupleDomainForLocalFilters` and some code could be reused between `DynamicFilterService` and `LocalDynamicFiltersCollector`

Reply:
I did a bit different refactoring - allowing a bit more efficient implementation of `DynamicFilter`, by caching the resulting predicate and the blocked future:
https://github.com/prestosql/presto/pull/3414/files#diff-fc014adc6614290fe6f8a48d7b0f6739
WDYT?